server/vehicle-persistence.lua

```diff
@@ -121,4 +118,135 @@ AddEventHandler('entityRemoved', function(entity)
         local passenger = passengers[i]
         SetPedIntoVehicle(passenger.ped, veh, passenger.seat)
     end
-end)
\ No newline at end of file
+end)
+
+if persistence == 'full' then return end
+
+local cachedVehicles = {}
+local config = require 'config.server'
+
+---@param plate string
+---@return boolean
+local function isVehicleSpawned(plate)
+    local vehicles = GetGamePool('CVehicle')
+
+    for i = 1, #vehicles do
+        local vehicle = vehicles[i]
+        if qbx.getVehiclePlate(vehicle) == plate then
+            return true
+        end
+    end
+
+    return false
+end
+
+--- Save the vehicle position to the database
+---@param vehicle number
+---@param coords vector3
+---@param heading number
+local function saveVehiclePosition(vehicle, coords, heading)
+    local type = GetVehicleType(vehicle)
+    if type == 'heli' or type == 'plane' then
+        coords = vec3(coords.x, coords.y, coords.z + 1.0)
+    end
+    exports.qbx_vehicles:SaveVehicle(vehicle, {
+        coords = vec4(coords.x, coords.y, coords.z, heading)
+    })
+end
+
+--- Save all vehicle positions to the database
+local function saveAllVehiclePosition()
+    local vehicles = GetGamePool('CVehicle')
+    for i = 1, #vehicles do
+        local vehicle = vehicles[i]
+        if DoesEntityExist(vehicle) and Entity(vehicle).state.persisted then
+            saveVehiclePosition(vehicle, GetEntityCoords(vehicle), GetEntityHeading(vehicle))
+        end
+    end
+end
+
+---@param coords vector4
+---@param id number
+---@param model string
+---@param props table
+local function spawnVehicle(coords, id, model, props)
+    if not coords or not id or not model or not props then return end
+
+    local _, veh = qbx.spawnVehicle({
+        spawnSource = vec4(coords.x, coords.y, coords.z, coords.w),
+        model = model,
+        props = props
+    })
+
+    cachedVehicles[id] = nil
+    Entity(veh).state:set('vehicleid', id, false)
+    TriggerClientEvent('qbx_core:client:removeVehZone', -1, id)
+    config.setVehicleLock(veh, config.persistence.lockState)
+end
+
+lib.callback.register('qbx_core:server:getVehiclesToSpawn', function()
+    return cachedVehicles
+end)
+
+AddEventHandler('onResourceStart', function(resourceName)
+    if resourceName ~= 'qbx_vehicles' then return end
+
+    local vehicles = exports.qbx_vehicles:GetPlayerVehicles({ states = 0 })
+    if not vehicles then return end
+
+    for i = 1, #vehicles do
+        local vehicle = vehicles[i]
+        if vehicle.coords and vehicle.props and vehicle.props.plate and
+            not isVehicleSpawned(vehicle.props.plate) then
```

Comment:
This being on the next line makes it look weird with the line below. Better to keep this on one line instead.